### PR TITLE
[rfw] Add `OverflowBar` widget and Update `ButtonBar` widget implementation

### PR DESCRIPTION
**Pull request by TahaTesser**
fixes [[`RFW`] Replace `ButtonBar` with a backwards compatible implementation using `OverflowBar`](https://github.com/flutter/flutter/issues/140381)

--- 

### Description

This adds  `OverflowBar` widget to material widgets and updates `ButtonBar` (to be deprecated widget) to use `OverflowBar` underneath for backwards compatibility.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by TahaTesser** [comment]
@Hixie 
There is  a tiny difference of `0.69%` when replacing `ButtonBar` with `OverflowBar`. The user can still call `OverflowBar` but it uses `OverflowBar` underneath.
```console
00:13 +79: /b/s/w/ir/x/w/packages/packages/rfw/test/material_widgets_test.dart: Material widgets                                                                                                       
══╡ EXCEPTION CAUGHT BY FLUTTER TEST FRAMEWORK ╞════════════════════════════════════════════════════
The following assertion was thrown while running async test code:
Golden "goldens/material_test.drawer.png": Pixel test failed, 0.69% diff detected.
```

And I also added `OverflowBar` widget separately for anyone wanting to use the correct widget moving forward. Its also produces a diff even tho this is a new widget and  a new test. I updated the golden files following the `rfw` contributing [documentation](OverflowBar) on updating golden files by running `run_tests.sh --update-goldens` script.


With the note above, marked this as ready to review and wait for your feedback.

File: packages/rfw/lib/src/flutter/material_widgets.dart
Comment (Hixie) [inline packages/rfw/lib/src/flutter/material_widgets.dart:122]: we should make it clear that the `ButtonBar` widget in `package:flutter/material.dart` is the one that's deprecated, and that the one here is never going to go away.

File: packages/rfw/lib/src/flutter/material_widgets.dart
Comment (Hixie) [inline packages/rfw/lib/src/flutter/material_widgets.dart:125]: this removes support for `buttonMinWidth` which violates our promise of keeping 100% backwards compatibility.

Comment (TahaTesser) [reply packages/rfw/lib/src/flutter/material_widgets.dart:125]: Apologies for not mentioning it in the description.

`buttonMinWidth` doesn't do anything since the implementation of new buttons based on `ButtonStyleButton`.   Please read Hans' comment [here](https://github.com/flutter/flutter/issues/79171#issuecomment-819695638) for more details.

The older Material buttons that used this property are no available in the framework. When this parameter was implemented, those older Material buttons used the provided value from `buttonMinWidth` parameter to adjust their min width.

File: packages/rfw/lib/src/flutter/material_widgets.dart
Comment (TahaTesser) [reply packages/rfw/lib/src/flutter/material_widgets.dart]: Thanks! Updated the docs.

**Comment by TahaTesser** [comment]
> I don't understand the diff, can you explain it in more detail? Why is the new rendering different?

The diff was produced when running `run_tests.sh --update-goldens` on my M1 MacBook.

Since the original goldens files are produced on Linux environment. 
https://github.com/flutter/packages/blob/0744fe6fdb31933d05d11699e7f44e1dd2c63e48/packages/rfw/test/utils.dart#L23

I've the updated the goldens from a linux environment, no diff is produced now for existing golden files. Only New golden file is added for the new widget test.

File: packages/rfw/lib/src/flutter/material_widgets.dart
Comment (Hixie) [reply packages/rfw/lib/src/flutter/material_widgets.dart:125]: let's at a minimum add some documentation to the big comment at the top of this function so that the API docs acknowledge that the ButtonBar widget has changed over time.

File: packages/rfw/test/material_widgets_test.dart
Comment (Hixie) [inline packages/rfw/test/material_widgets_test.dart:209]: learning from my mistake with ButtonBar, could we add more variants of this test with different argument values and so on?

File: packages/rfw/lib/src/flutter/material_widgets.dart
Comment (TahaTesser) [reply packages/rfw/lib/src/flutter/material_widgets.dart:125]: `buttonMinWidth`, `buttonHeight`, and `buttonAlignedDropdown` are no longer supported by widgets in the framework. I added some documentation in the comment above.

File: packages/rfw/test/material_widgets_test.dart
Comment (TahaTesser) [reply packages/rfw/test/material_widgets_test.dart:260]: I just added another test for `OverflowBar` properties. This includes all the `OverflowBar` and tests in the regular  and overflow state.

When writing this test I found out `OverflowBar.clipBehavior` is simply passed but not used in the `OverflowBar` implementation. I'll file a bug for this and fix it.

Comment (TahaTesser) [reply packages/rfw/test/material_widgets_test.dart:260]: I also did a clean up to reduce code duplication.

File: packages/rfw/lib/src/flutter/material_widgets.dart
Comment (Hixie) [inline packages/rfw/lib/src/flutter/material_widgets.dart:122]: "note that" -> https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-empty-prose

File: packages/rfw/lib/src/flutter/material_widgets.dart
Comment (Hixie) [inline packages/rfw/lib/src/flutter/material_widgets.dart:126]: "not" -> "no longer" ?

File: packages/rfw/lib/src/flutter/material_widgets.dart
Comment (Hixie) [reply packages/rfw/lib/src/flutter/material_widgets.dart:125]: That comment is good. I was thinking more of the API documentation around line 52, though (maybe a new paragraph that says something about like "Some features have changed in the underlying Material library and are therefore no longer supported, including:" followed by a description of what's happened with `ButtonBar`).

Comment (TahaTesser) [reply packages/rfw/lib/src/flutter/material_widgets.dart:125]: Sounds good.

File: packages/rfw/lib/src/flutter/material_widgets.dart
Comment (Hixie) [inline packages/rfw/lib/src/flutter/material_widgets.dart:66]: we should use markdown in this section

Comment (TahaTesser) [reply packages/rfw/lib/src/flutter/material_widgets.dart]: Oops. Good catch. i also fixed the format.

**Comment by TahaTesser** [comment]
@Hixie Resolved version conflicts.

**Comment by TahaTesser** [comment]
Resolved new conflicts

File: packages/rfw/lib/src/flutter/material_widgets.dart
Comment (Hixie) [inline packages/rfw/lib/src/flutter/material_widgets.dart:69]: this all needs to be markdown with hyperlinks and so on (sorry if i wasn't clear in the earlier comment).

What I mean is e.g. `ButtonBar` needs to be `[ButtonBar]`, `package:flutter/material.dart` needs to be whatever the magic incantation is to reference the `material` library (maybe just `[material]`?), `rfw` needs to either be in backticks or uppercase, the parameters in the list should be in backticks, etc.

**Review by Hixie — APPROVED** [review]
LGTM modulo the documentation fixes requested above

File: packages/rfw/lib/src/flutter/material_widgets.dart
Comment (TahaTesser) [reply packages/rfw/lib/src/flutter/material_widgets.dart:69]: Ahh my bad. Thanks for the suggestion!

`package:flutter/material.dart` can be simply referenced as "Flutter's material library". We reference like this in the framework too in various class.

**Comment by Hixie** [comment]
Looks like the test coverage is no longer at 100% since this PR.

**Comment by TahaTesser** [comment]
> Looks like the test coverage is no longer at 100% since this PR.

Filed a fix https://github.com/flutter/flutter/issues/142602

**Pull request merged**
